### PR TITLE
Use observations shape to estimate bins in posterior mean plot

### PR DESCRIPTION
**Pull request by RaulPL**
It changes how the number of bins is assigned for the posterior predictive mean plot

### CODE REVIEW

**Comment by aloctavodia** [comment]
Hi Raul, thanks for your contribution can we use `get_bins` that is inside `arviz/plots/plot_utils.py`

**Comment by RaulPL** [comment]
Yes, I will change the lines that define the number of bins like this:
```
nbins = round(len(obs_vals) ** 0.5)
```
to something like this:
```
nbins = len(get_bins(obs_vals))
```

**Comment by OriolAbril** [comment]
Hi, I have not had time to review everything in depth yet, only from my mobile, I hope my comment makes sense.

I think the number of bins _sould_ not have any effect on the normaliztio/y_scale. So much variation depending on the number of bins looks really weird to me, given that `density=True` is being used.

**Comment by ColCarroll** [comment]
Fixes #751 

@OriolAbril I'm not sure I understand your comment -- you saw the current behavior that this fixes? (I just now linked the issue!)

image

**Comment by OriolAbril** [comment]
Yes, comparing with the plot in the comment, why is the posterior predictive mean area seems to change? Are there too many bins and some of them end up empty so the real area differs from what seems colored?

**Comment by ColCarroll** [comment]
Oh I see -- I think the issue was that the current behavior plots 1 histogram for each chain, so the total area ended up being 4 (the number of chains in the rugby example).

**Comment by OriolAbril** [comment]
Oh, okey, that makes sense too. I could not properly read the code so I thought that the only change was the number of bins amd I was having trouble understanding how this change solved the issue. Sorry about that.

**Comment by ahartikainen** [comment]
I think it still should use the length of vals, not obs. The missing component was the get_bins function.

Integers and histograms fail when bin width is less than 1.

**Comment by ahartikainen** [comment]
The other option is that we could a scheme where the obs and mean pp uses the same bin_edges, then the histograms can be compared more easily.

This would need careful set-up, because in the worst case pp can be far from the obs (e.g. obs are near 0, pp are near 1e6)

**Comment by RaulPL** [comment]
Yes, using vals in get_bins also work:

image

File: arviz/plots/ppcplot.py [hunk 1]
```diff
@@ -401,7 +402,7 @@ def plot_ppc(
                     )
                 else:
                     vals = pp_vals.flatten()
-                    nbins = round(len(vals) ** 0.5)
+                    nbins = round(len(obs_vals) ** 0.5)
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/ppcplot.py]: Or is this for floats? (On mobile, so reading source is a bit of challenge)

**Comment by OriolAbril** [comment]
Hi Raul, first of all, thanks for reporting the issue and for working on this PR. 

Secondly, sorry for all the discussion about your PR without having properly read all your code, I hope the reasons why are clear, but just in case, I will try to explain a little the situation. The issue you found is due to using the wrong number of bins. We are working on discrete data (histograms are the default for discrete data and kde are the default for real data in ArviZ), therefore, having a bin width smaller than 1 makes no sense:

image

As you can see, the original behaviour is wrong because there are so many bins that the bin width is smaller than 1, therefore, the height of the bin has to be larger than it should be. To use a simple example, if we were using as bins `0.75, 1.25, 1.75, 2.25,...` as bins for discrete data, the bin `(1.25, 1.75)` would be empty, and to keep the area of the histogram equal to 1, the `(0.75, 1.25)` bin's height would double (with respect to using 1 as bin width -> `bin_edges=(0.75, 1.75)`.

In this example, using the number of observed values to get the number of bins solved the issue, but unfortunately, it was not a general solution, and the same issue of having too many bins -> bin width being too small could happen anyway. For instance, if the example data was about football, where the number of goals scored per match is much smaller than the number of matches in the league, `nbins =round(len(obs_vals) ** 0.5)` could still yield bin width smaller than 1. 

The root of the problem comes from using `nbins = round(len(array) ** 0.5)`. This is why `get_bins` should be used, because it takes into account the fact that data is discrete, the size of the array and its dispersion to return well behaved `bin_edges`. Therefore, to properly solve the issue once and for all, the code should be modified to always use `bins = get_bins(vals)` when performing a histogram of `vals` (or the same with `obs_vals` when performing a histogram of `obs_vals`). As `get_bins` already returns the bin edges, there is no need to use its length and then let numpy recalculate the bin edges.

File: arviz/plots/ppcplot.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (OriolAbril) [reply arviz/plots/ppcplot.py]: Histograms are always (and only) used for discrete data.

**Comment by RaulPL** [comment]
OK, now np.histogram is using the bins returned by get_bins:
image

**Comment by ahartikainen** [comment]
LGTM

**Comment by OriolAbril** [comment]
Looks great, thanks!

**Comment by ahartikainen** [comment]
Thanks for the PR

**Pull request merged**